lib/srv/regular/sshserver_test.go

```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+	c.Assert(err, IsNil)
+
+	// Create host key and certificate for node.
+	keys, err := s.server.Auth().GenerateServerKeys(auth.GenerateServerKeysRequest{
+		HostID:               "raw-node",
+		NodeName:             "raw-node",
+		Roles:                teleport.Roles{teleport.RoleNode},
+		AdditionalPrincipals: []string{hostname},
+		DNSNames:             []string{hostname},
+	})
+	c.Assert(err, IsNil)
+
+	signer, err := sshutils.NewSigner(keys.Key, keys.Cert)
+	c.Assert(err, IsNil)
+
+	// configure a server which allows any client to connect
+	cfg := ssh.ServerConfig{
+		NoClientAuth: true,
+		PasswordCallback: func(conn ssh.ConnMetadata, password []byte) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+	}
+	cfg.AddHostKey(signer)
+
+	listener, err := net.Listen("tcp", ":0")
+	c.Assert(err, IsNil)
+
+	_, port, err := net.SplitHostPort(listener.Addr().String())
+	c.Assert(err, IsNil)
+
+	addr := net.JoinHostPort(hostname, port)
+
+	return &rawNode{
+		listener: listener,
+		cfg:      cfg,
+		addr:     addr,
+	}
+}
+
+// startX11EchoServer starts a fake node which, for each incomging SSH connection, accepts an
+// X11 forwarding request and then dials a single X11 channel which echoes all bytes written
+// to it.  Used to verify the behavior of X11 forwarding in recording proxies.
+func (s *SrvSuite) startX11EchoServer(ctx context.Context, c *C) *rawNode {
+	node := s.newRawNode(c)
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		defer cancel()
+		for {
+			conn, chs, _, err := node.accept()
+			if err != nil {
+				log.Warnf("X11 echo server closing: %v", err)
+				return
+			}
+			go func() {
+				defer conn.Close()
+
+				// expect client to open a session channel
+				var nch ssh.NewChannel
+				select {
+				case nch = <-chs:
+				case <-ctx.Done():
+					return
+				}
+				c.Assert(nch.ChannelType(), Equals, teleport.ChanSession)
+
+				_, creqs, err := nch.Accept()
+				c.Assert(err, IsNil)
+
+				// expect client to send an X11 forwarding request
+				var req *ssh.Request
+				select {
+				case req = <-creqs:
+				case <-ctx.Done():
+					return
+				}
+
+				c.Assert(req.Type, Equals, sshutils.X11ForwardRequest)
+				c.Assert(req.Reply(true, nil), IsNil)
+
+				// start a fake X11 channel
+				xch, _, err := conn.OpenChannel(sshutils.X11ChannelRequest, nil)
+				c.Assert(err, IsNil)
+
+				defer xch.Close()
+				// echo all bytes back across the X11 channel
+				if _, err := io.Copy(xch, xch); err != nil {
+					if err == io.EOF {
+						xch.CloseWrite()
+					} else {
+						log.Errorf("X11 channel error: %v", err)
+					}
+				}
+			}()
+		}
+	}()
+	return node
+}
+
+func writeAll(dst io.Writer, buf []byte) error {
+	w := 0
+	for w < len(buf) {
+		n, err := dst.Write(buf[w:])
+		if err != nil {
+			return err
+		}
+		w += n
+	}
+	return nil
+}
```

Comment:
`io.Writer` should always return a non-nil error on a short write.
this helper func likely behaves just like `dst.Write(buf)`

Reply:
Woah, I always assumed it returned after the first successful write... cool!